errai-cdi/errai-cdi-client/src/main/java/org/jboss/errai/enterprise/client/cdi/InstanceProvider.java

```diff
@@ -90,7 +90,11 @@ public boolean isAmbiguous() {
 
     @Override
     public Iterator<Object> iterator() {
-      return Collections.emptyList().iterator();
+      Collection beanDefs = IOC.getBeanManager().lookupBeans(type, qualifiers);
+      if(beanDefs==null){
+        return Collections.emptyList().iterator();
+      }
+      return beanDefs.iterator();
```

Comment:
This line returns an `Iterator<SyncBeanDef>`, but what we want here is an iterator of instances (i.e. `Iterator<T>`).